src/uct/cuda/cuda_ipc/cuda_ipc_iface.c

```diff
@@ -343,22 +338,33 @@ static ucs_mpool_ops_t uct_cuda_ipc_event_desc_mpool_ops = {
     .obj_cleanup   = uct_cuda_ipc_event_desc_cleanup,
 };
 
+ucs_status_t uct_cuda_ipc_unmap_memhandle(void *mapped_addr)
+{
+    return UCT_CUDADRV_FUNC(cuIpcCloseMemHandle((CUdeviceptr)mapped_addr));
+}
+
 ucs_status_t uct_cuda_ipc_map_memhandle(void *arg, uct_cuda_ipc_key_t *key,
                                         void **mapped_addr)
 {
+    if (key->unmap_memhandle_func == 0) {
+        key->unmap_memhandle_func = uct_cuda_ipc_unmap_memhandle;
+    }
+
     if (key->d_mapped != 0) {
         /* potentially already mapped in uct_cuda_ipc_rkey_unpack */
         *mapped_addr = (void *) key->d_mapped;
         return UCS_OK;
     }
 
-    return  UCT_CUDADRV_FUNC(cuIpcOpenMemHandle((CUdeviceptr *)mapped_addr,
-                             key->ph, CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS));
-}
+    ucs_status_t status = UCT_CUDADRV_FUNC(cuIpcOpenMemHandle((CUdeviceptr *)mapped_addr,
+                                           key->ph, CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS));
 
-ucs_status_t uct_cuda_ipc_unmap_memhandle(void *mapped_addr)
-{
-    return UCT_CUDADRV_FUNC(cuIpcCloseMemHandle((CUdeviceptr)mapped_addr));
+    /* Store mapped memory handle to use when copying remaining blocks */
+    if (status == UCS_OK && key->d_mapped == 0) {
```

Comment:
code style: please use `()` for each statement. 
`((status == UCS_OK) && (key->d_mapped == 0))`